unstructured/partition/xml.py

```diff
@@ -20,41 +22,58 @@
 from unstructured.partition.text import element_from_text
 
 
-def is_leaf(elem):
-    return not bool(elem)
-
-
-def is_string(elem):
-    return isinstance(elem, str) or (hasattr(elem, "text") and isinstance(elem.text, str))
-
-
 def get_leaf_elements(
     filename: Optional[str] = None,
     file: Optional[Union[IO[bytes], SpooledTemporaryFile]] = None,
     text: Optional[str] = None,
-    xml_path: str = ".",
-    xml_keep_tags: bool = False,
+    xml_path: Optional[str] = None,
 ) -> List[Optional[str]]:
+    """Get leaf elements from the XML tree defined in filename, file, or text."""
     exactly_one(filename=filename, file=file, text=text)
     if filename:
-        _, raw_text = read_txt_file(filename=filename)
+        return _get_leaf_elements(filename, xml_path=xml_path)
     elif file:
-        f = spooled_to_bytes_io_if_needed(
-            cast(Union[BinaryIO, SpooledTemporaryFile], file),
+        f = cast(
+            IO[bytes],
+            spooled_to_bytes_io_if_needed(
+                cast(Union[BinaryIO, SpooledTemporaryFile], file),
+            ),
         )
-        _, raw_text = read_txt_file(file=f)
-    elif text:
-        raw_text = text
-
-    root = ET.fromstring(raw_text)
-    leaf_elements = []
+        return _get_leaf_elements(f, xml_path=xml_path)
+    else:
+        b = BytesIO(bytes(cast(str, text), encoding="utf-8"))
+        return _get_leaf_elements(b, xml_path=xml_path)
 
-    for elem in root.findall(xml_path):
-        for subelem in elem.iter():
-            if is_leaf(subelem) and is_string(subelem.text):
-                leaf_elements.append(subelem.text)
 
-    return leaf_elements
+def _get_leaf_elements(
```

Comment:
is the return ever used as a list or just used as iterators? If it is only used as iterator later might be even better to return a generator instead of list (so that even more efficient for memory)

Reply:
That's a good point, I'll check the downstream uses. My only concern would be if we actually use it lazily then I assume the file will stay open, but in practice that probably won't matter much.

Reply:
So the immediate thing is that having `get_leaf_elements` and `_get_leaf_elements` return a generator doesn't have much affect since `partition_xml` immediately wraps it in a list. If we go all the way and have `partition_xml` return a generator, then a number of tests fail.

Tests can be updated, but since `partition_pdf` is at the boundary of our interface, I feel like we should hold off on such a change unless we want to make that change consistent across all `partition` functions.

I'll make `get_leaf_elements` and `_get_leaf_elements` return generators though.